File: app/src/main/java/com/dmitrybrant/android/mandelbrot/MandelbrotViewBase.java
package com.dmitrybrant.android.mandelbrot;

import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.List;

import android.graphics.PointF;
import androidx.annotation.NonNull;
import android.text.format.DateUtils;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.view.MotionEvent;

public abstract class MandelbrotViewBase extends View {
    private static final String TAG = "MandelbrotViewBase";
    public static final int DEFAULT_ITERATIONS = 128;
    public static final int MAX_ITERATIONS = 2048;
    public static final int MIN_ITERATIONS = 2;
    public static final double DEFAULT_X_CENTER = -0.5;
    public static final double DEFAULT_Y_CENTER = 0.0;
    public static final double DEFAULT_X_EXTENT = 3.0;
    public static final double DEFAULT_JULIA_X_CENTER = 0.0;
    public static final double DEFAULT_JULIA_Y_CENTER = 0.0;
    public static final double DEFAULT_JULIA_EXTENT = 3.0;

    private static final int TOUCH_NONE = 0;
    private static final int TOUCH_ROTATE = 1;
    private static final int TOUCH_ZOOM = 2;

    private static final float CROSSHAIR_WIDTH = 16f;

    private boolean isJulia;
    private int paramIndex;

    private List<Thread> currentThreads = new ArrayList<>();
    private volatile boolean terminateThreads;

    private Paint paint;
    private Bitmap viewportBitmap;
    private Rect viewportRect;
    private boolean showCrosshairs;

    private int screenWidth;
    private int screenHeight;

    private double xcenter;
    public double getXCenter() { return xcenter; }
    public void setXCenter(double xcenter) { this.xcenter = xcenter; }

    private double ycenter;
    public double getYCenter() { return ycenter; }
    public void setYCenter(double ycenter) { this.ycenter = ycenter; }

    private double xextent;
    public double getXExtent() { return xextent; }
    public void setXExtent(double xextent) { this.xextent = xextent; }

    private double xmin;
    private double xmax;
    private double ymin;
    private double ymax;
    private double jx;
    private double jy;

    private int numIterations = DEFAULT_ITERATIONS;
    public int getNumIterations() { return numIterations; }
    public void setNumIterations(int iter) {
        numIterations = iter;
        if (numIterations < MIN_ITERATIONS) {
            numIterations = MIN_ITERATIONS;
        }
        if (numIterations > MAX_ITERATIONS) {
            numIterations = MAX_ITERATIONS;
        }
    }

    private int startCoarseness = 16;
    private int endCoarseness = 1;

    private float previousX;
    private float previousY;
    private PointF pinchStartPoint = new PointF();
    private float pinchStartDistance = 0.0f;
    private int touchMode = TOUCH_NONE;

    private float displayDensity;

    public interface OnPointSelected {
        void pointSelected(double x, double y);
    }
    private OnPointSelected onPointSelected;
    public void setOnPointSelected(OnPointSelected listener) {
        onPointSelected = listener;
    }

    public interface OnCoordinatesChanged {
        void newCoordinates(double xmin, double xmax, double ymin, double ymax);
    }
    private OnCoordinatesChanged onCoordinatesChanged;
    public void setOnCoordinatesChanged(OnCoordinatesChanged listener) {
        onCoordinatesChanged = listener;
    }

    public MandelbrotViewBase(Context context) {
        super(context);
    }

    public MandelbrotViewBase(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public MandelbrotViewBase(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
    }

    protected void init(boolean isJulia) {
        if (isInEditMode()) {
            return;
        }
        this.isJulia = isJulia;
        this.paramIndex = isJulia ? 1 : 0;
        displayDensity = getResources().getDisplayMetrics().density;

        paint = new Paint();
        paint.setStyle(Paint.Style.FILL);
        paint.setColor(Color.WHITE);
        paint.setStrokeWidth(1.5f * displayDensity);
    }

    @Override
    public void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        if (isInEditMode()) {
            return;
        }
        if ((screenWidth != this.getWidth()) || (screenHeight != this.getHeight())) {
            screenWidth = this.getWidth();
            screenHeight = this.getHeight();
            if ((screenWidth == 0) || (screenHeight == 0)) {
                return;
            }

            terminateThreads();
            initMinMax();

            viewportBitmap = Bitmap.createBitmap(screenWidth, screenHeight, Bitmap.Config.ARGB_8888);
            mandelnative.SetBitmap(paramIndex, viewportBitmap);
            viewportRect = new Rect(0, 0, viewportBitmap.getWidth() - 1, viewportBitmap.getHeight() - 1);

            render();
            return;
        }
        if ((screenWidth == 0) || (screenHeight == 0)) {
            return;
        }

        mandelnative.UpdateBitmap(paramIndex, viewportBitmap);
        canvas.drawBitmap(viewportBitmap, viewportRect, viewportRect, paint);

        if (showCrosshairs) {
            canvas.drawLine(screenWidth / 2 - CROSSHAIR_WIDTH * displayDensity, screenHeight / 2, screenWidth / 2 + CROSSHAIR_WIDTH * displayDensity, screenHeight / 2, paint);
            canvas.drawLine(screenWidth / 2, screenHeight / 2 - CROSSHAIR_WIDTH * displayDensity, screenWidth / 2, screenHeight / 2 + CROSSHAIR_WIDTH * displayDensity, paint);
        }
    }

    @Override
    public boolean onTouchEvent(@NonNull MotionEvent event){
        switch (event.getAction() & MotionEvent.ACTION_MASK) {
            case MotionEvent.ACTION_DOWN:
                previousX = event.getX();
                previousY = event.getY();
                endCoarseness = startCoarseness;
                break;

            case MotionEvent.ACTION_MOVE:
                if (event.getPointerCount() == 1) {
                    if (touchMode != TOUCH_ROTATE) {
                        previousX = event.getX();
                        previousY = event.getY();
                    }
                    touchMode = TOUCH_ROTATE;
                    float x = event.getX();
                    float y = event.getY();
                    float dx = x - previousX;
                    float dy = y - previousY;
                    previousX = x;
                    previousY = y;

                    if((dx != 0) || (dy != 0)){
                        double amountX = ((double)dx / (double)screenWidth) * (xmax - xmin);
                        double amountY = ((double)dy / (double)screenHeight) * (ymax - ymin);
                        xmin -= amountX; xmax -= amountX;
                        ymin -= amountY; ymax -= amountY;
                    }

                } else if (event.getPointerCount() == 2) {
                    if (touchMode != TOUCH_ZOOM) {
                        pinchStartDistance = getPinchDistance(event);
                        getPinchCenterPoint(event, pinchStartPoint);
                        previousX = pinchStartPoint.x;
                        previousY = pinchStartPoint.y;
                        touchMode = TOUCH_ZOOM;
                    } else {
                        PointF pt = new PointF();
                        getPinchCenterPoint(event, pt);
                        previousX = pt.x;
                        previousY = pt.y;
                        float pinchScale = getPinchDistance(event) / pinchStartDistance;
                        pinchStartDistance = getPinchDistance(event);

                        if (pinchScale > 0) {
                            double xCenter = xmin + ((xmax - xmin) * ((double) pt.x / screenWidth));
                            double yCenter = ymin + ((ymax - ymin) * ((double) pt.y / screenHeight));
                            xmin = xCenter - (xCenter - xmin) / pinchScale;
                            xmax = xCenter + (xmax - xCenter) / pinchScale;
                            ymin = yCenter - (yCenter - ymin) / pinchScale;
                            ymax = yCenter + (ymax - yCenter) / pinchScale;
                        }
                    }
                }
                if (onPointSelected != null) {
                    onPointSelected.pointSelected(xmin + ((double)event.getX() * (xmax - xmin) / screenWidth),
                            ymin + ((double)event.getY() * (ymax - ymin) / screenHeight));
                }
                endCoarseness = startCoarseness;
                break;

            case MotionEvent.ACTION_UP:
            case MotionEvent.ACTION_CANCEL:
                pinchStartPoint.x = 0.0f;
                pinchStartPoint.y = 0.0f;
                touchMode = TOUCH_NONE;
                endCoarseness = 1;
                if (onPointSelected != null) {
                    onPointSelected.pointSelected(xmin + ((double)event.getX() * (xmax - xmin) / screenWidth),
                            ymin + ((double)event.getY() * (ymax - ymin) / screenHeight));
                }
                break;
        }

        render();
        return true;
    }

    public void requestCoordinates() {
        if (onCoordinatesChanged != null) {
            onCoordinatesChanged.newCoordinates(xmin, xmax, ymin, ymax);
        }
    }

    public void render() {
        terminateThreads();
        if (getVisibility() != View.VISIBLE) {
            return;
        }
        if (onCoordinatesChanged != null) {
            onCoordinatesChanged.newCoordinates(xmin, xmax, ymin, ymax);
        }

        xextent = xmax - xmin;
        xcenter = xmin + xextent / 2.0;
        ycenter = ymin + (ymax - ymin) / 2.0;

        mandelnative.SetParameters(paramIndex, numIterations, xmin, xmax, ymin, ymax,
                isJulia ? 1 : 0, jx, jy, screenWidth, screenHeight);
        Thread t;

        t = new MandelThread(0, 0, screenWidth, screenHeight / 2, startCoarseness);
        t.start();
        currentThreads.add(t);
        t = new MandelThread(0, screenHeight / 2, screenWidth, screenHeight / 2, startCoarseness);
        t.start();
        currentThreads.add(t);
    }

    public void terminateThreads() {
        try {
            mandelnative.SignalTerminate(paramIndex);
            terminateThreads = true;

            for(Thread t : currentThreads){
                if (t.isAlive()) {
                    t.join(DateUtils.SECOND_IN_MILLIS);
                }
                if (t.isAlive()) {
                    Log.w(TAG, "Thread is still alive after 1sec...");
                }
            }

            terminateThreads = false;
            currentThreads.clear();
        } catch(Exception ex) {
            Log.w(TAG, "Exception while terminating threads: " + ex.getMessage());
        }
    }

    public void reset() {
        if (isJulia) {
            xcenter = DEFAULT_JULIA_X_CENTER;
            ycenter = DEFAULT_JULIA_Y_CENTER;
            xextent = DEFAULT_JULIA_EXTENT;
        } else {
            xcenter = DEFAULT_X_CENTER;
            ycenter = DEFAULT_Y_CENTER;
            xextent = DEFAULT_X_EXTENT;
        }
        numIterations = DEFAULT_ITERATIONS;
        initMinMax();
        render();
    }

    public void setColorScheme(int[] colors) {
        mandelnative.SetColorPalette(paramIndex, colors, colors.length);
    }

    public void setJuliaCoords(double jx, double jy) {
        this.jx = jx;
        this.jy = jy;
    }

    public void setCrosshairsEnabled(boolean enabled) {
        showCrosshairs = enabled;
    }

    public void savePicture(@NonNull String fileName) throws IOException
    {
        savePicture(new FileOutputStream(fileName));
    }

    public void savePicture(@NonNull OutputStream stream) throws IOException
    {
        viewportBitmap.compress(Bitmap.CompressFormat.PNG, 100, stream);
        stream.flush();
        stream.close();
    }

    private void initMinMax() {
        double ratio = (double)screenHeight / (double)(screenWidth == 0 ? 1 : screenWidth);
        xmin = xcenter - xextent / 2.0;
        xmax = xcenter + xextent / 2.0;
        ymin = ycenter - ratio * xextent / 2.0;
        ymax = ycenter + ratio * xextent / 2.0;
    }

    private float getPinchDistance(MotionEvent event) {
        float x = event.getX(0) - event.getX(1);
        float y = event.getY(0) - event.getY(1);
        return (float) Math.sqrt(x * x + y * y);
    }

    private void getPinchCenterPoint(MotionEvent event, PointF pt) {
        pt.x = (event.getX(0) + event.getX(1)) * 0.5f;
        pt.y = (event.getY(0) + event.getY(1)) * 0.5f;
    }

    private class MandelThread extends Thread
    {
        MandelThread (int x, int y, int width, int height, int level) {
            startX = x;
            startY = y;
            startWidth = width;
            startHeight = height;
            this.level = level;
        }

        private int startX, startY;
        private int startWidth, startHeight;
        private int level;

        @Override
        public void run(){
            int curLevel = level;
            while (true) {
                mandelnative.DrawFractal(paramIndex, startX, startY, startWidth, startHeight, curLevel, curLevel == level ? 1 : 0);
                MandelbrotViewBase.this.postInvalidate();
                if (terminateThreads) {
                    break;
                }
                if (curLevel <= endCoarseness) {
                    break;
                }
                curLevel /= 2;
            }
        }    
    }
}

File: app/src/main/java/com/dmitrybrant/android/mandelbrot/MandelbrotViewBase.kt
package com.dmitrybrant.android.mandelbrot

import android.content.Context
import android.graphics.*
import android.text.format.DateUtils
import android.util.AttributeSet
import android.util.Log
import android.view.MotionEvent
import android.view.View
import java.io.FileOutputStream
import java.io.IOException
import java.io.OutputStream
import java.util.*
import kotlin.math.sqrt

abstract class MandelbrotViewBase @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0)
    : View(context, attrs, defStyleAttr) {

    interface OnPointSelected {
        fun pointSelected(x: Double, y: Double)
    }

    interface OnCoordinatesChanged {
        fun newCoordinates(xmin: Double, xmax: Double, ymin: Double, ymax: Double)
    }

    private var isJulia = false
    private var paramIndex = 0
    private val currentThreads: MutableList<Thread> = ArrayList()
    @Volatile private var terminateThreads = false
    private var paint: Paint? = null
    private var viewportBitmap: Bitmap? = null
    private var viewportRect: Rect? = null
    private var showCrosshairs = false
    private var screenWidth = 0
    private var screenHeight = 0

    var xCenter = 0.0
    var yCenter = 0.0
    var xExtent = 0.0

    private var xmin = 0.0
    private var xmax = 0.0
    private var ymin = 0.0
    private var ymax = 0.0
    private var jx = 0.0
    private var jy = 0.0

    private var numIterations = DEFAULT_ITERATIONS

    fun getNumIterations(): Int {
        return numIterations
    }

    fun setNumIterations(iter: Int) {
        numIterations = iter
        if (numIterations < MIN_ITERATIONS) {
            numIterations = MIN_ITERATIONS
        }
        if (numIterations > MAX_ITERATIONS) {
            numIterations = MAX_ITERATIONS
        }
    }

    private val startCoarseness = 16
    private var endCoarseness = 1
    private var previousX = 0f
    private var previousY = 0f
    private val pinchStartPoint = PointF()
    private var pinchStartDistance = 0.0f
    private var touchMode = TOUCH_NONE
    private var displayDensity = 0f

    private var onPointSelected: OnPointSelected? = null
    fun setOnPointSelected(listener: OnPointSelected) {
        onPointSelected = listener
    }

    private var onCoordinatesChanged: OnCoordinatesChanged? = null
    fun setOnCoordinatesChanged(listener: OnCoordinatesChanged) {
        onCoordinatesChanged = listener
    }

    protected fun setup(isJulia: Boolean) {
        if (isInEditMode) {
            return
        }
        this.isJulia = isJulia
        paramIndex = if (isJulia) 1 else 0
        displayDensity = resources.displayMetrics.density
        paint = Paint()
        paint!!.style = Paint.Style.FILL
        paint!!.color = Color.WHITE
        paint!!.strokeWidth = 1.5f * displayDensity
    }

    public override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        if (isInEditMode) {
            return
        }
        if (screenWidth != this.width || screenHeight != this.height) {
            screenWidth = this.width
            screenHeight = this.height
            if (screenWidth == 0 || screenHeight == 0) {
                return
            }
            terminateThreads()
            initMinMax()
            viewportBitmap = Bitmap.createBitmap(screenWidth, screenHeight, Bitmap.Config.ARGB_8888)
            MandelNative.setBitmap(paramIndex, viewportBitmap)
            viewportRect = Rect(0, 0, viewportBitmap!!.width - 1, viewportBitmap!!.height - 1)
            render()
            return
        }
        if (screenWidth == 0 || screenHeight == 0) {
            return
        }
        MandelNative.updateBitmap(paramIndex, viewportBitmap)
        canvas.drawBitmap(viewportBitmap!!, viewportRect, viewportRect!!, paint)
        if (showCrosshairs) {
            canvas.drawLine(screenWidth / 2 - CROSSHAIR_WIDTH * displayDensity, screenHeight / 2.toFloat(), screenWidth / 2 + CROSSHAIR_WIDTH * displayDensity, screenHeight / 2.toFloat(), paint!!)
            canvas.drawLine(screenWidth / 2.toFloat(), screenHeight / 2 - CROSSHAIR_WIDTH * displayDensity, screenWidth / 2.toFloat(), screenHeight / 2 + CROSSHAIR_WIDTH * displayDensity, paint!!)
        }
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        when (event.action and MotionEvent.ACTION_MASK) {
            MotionEvent.ACTION_DOWN -> {
                previousX = event.x
                previousY = event.y
                endCoarseness = startCoarseness
            }
            MotionEvent.ACTION_MOVE -> {
                if (event.pointerCount == 1) {
                    if (touchMode != TOUCH_ROTATE) {
                        previousX = event.x
                        previousY = event.y
                    }
                    touchMode = TOUCH_ROTATE
                    val x = event.x
                    val y = event.y
                    val dx = x - previousX
                    val dy = y - previousY
                    previousX = x
                    previousY = y
                    if (dx != 0f || dy != 0f) {
                        val amountX = dx.toDouble() / screenWidth.toDouble() * (xmax - xmin)
                        val amountY = dy.toDouble() / screenHeight.toDouble() * (ymax - ymin)
                        xmin -= amountX
                        xmax -= amountX
                        ymin -= amountY
                        ymax -= amountY
                    }
                } else if (event.pointerCount == 2) {
                    if (touchMode != TOUCH_ZOOM) {
                        pinchStartDistance = getPinchDistance(event)
                        getPinchCenterPoint(event, pinchStartPoint)
                        previousX = pinchStartPoint.x
                        previousY = pinchStartPoint.y
                        touchMode = TOUCH_ZOOM
                    } else {
                        val pt = PointF()
                        getPinchCenterPoint(event, pt)
                        previousX = pt.x
                        previousY = pt.y
                        val pinchScale = getPinchDistance(event) / pinchStartDistance
                        pinchStartDistance = getPinchDistance(event)
                        if (pinchScale > 0) {
                            val xCenter = xmin + (xmax - xmin) * (pt.x.toDouble() / screenWidth)
                            val yCenter = ymin + (ymax - ymin) * (pt.y.toDouble() / screenHeight)
                            xmin = xCenter - (xCenter - xmin) / pinchScale
                            xmax = xCenter + (xmax - xCenter) / pinchScale
                            ymin = yCenter - (yCenter - ymin) / pinchScale
                            ymax = yCenter + (ymax - yCenter) / pinchScale
                        }
                    }
                }
                if (onPointSelected != null) {
                    onPointSelected!!.pointSelected(xmin + event.x.toDouble() * (xmax - xmin) / screenWidth,
                            ymin + event.y.toDouble() * (ymax - ymin) / screenHeight)
                }
                endCoarseness = startCoarseness
            }
            MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
                pinchStartPoint.x = 0.0f
                pinchStartPoint.y = 0.0f
                touchMode = TOUCH_NONE
                endCoarseness = 1
                if (onPointSelected != null) {
                    onPointSelected!!.pointSelected(xmin + event.x.toDouble() * (xmax - xmin) / screenWidth,
                            ymin + event.y.toDouble() * (ymax - ymin) / screenHeight)
                }
            }
        }
        render()
        return true
    }

    fun requestCoordinates() {
        if (onCoordinatesChanged != null) {
            onCoordinatesChanged!!.newCoordinates(xmin, xmax, ymin, ymax)
        }
    }

    fun render() {
        terminateThreads()
        if (visibility != VISIBLE) {
            return
        }
        if (onCoordinatesChanged != null) {
            onCoordinatesChanged!!.newCoordinates(xmin, xmax, ymin, ymax)
        }
        xExtent = xmax - xmin
        xCenter = xmin + xExtent / 2.0
        yCenter = ymin + (ymax - ymin) / 2.0
        MandelNative.setParameters(paramIndex, numIterations, xmin, xmax, ymin, ymax,
                if (isJulia) 1 else 0, jx, jy, screenWidth, screenHeight)
        var t: Thread
        t = MandelThread(0, 0, screenWidth, screenHeight / 2, startCoarseness)
        t.start()
        currentThreads.add(t)
        t = MandelThread(0, screenHeight / 2, screenWidth, screenHeight / 2, startCoarseness)
        t.start()
        currentThreads.add(t)
    }

    fun terminateThreads() {
        try {
            MandelNative.signalTerminate(paramIndex)
            terminateThreads = true
            for (t in currentThreads) {
                if (t.isAlive) {
                    t.join(DateUtils.SECOND_IN_MILLIS)
                }
                if (t.isAlive) {
                    Log.w(TAG, "Thread is still alive after 1sec...")
                }
            }
            terminateThreads = false
            currentThreads.clear()
        } catch (ex: Exception) {
            Log.w(TAG, "Exception while terminating threads: " + ex.message)
        }
    }

    fun reset() {
        if (isJulia) {
            xCenter = DEFAULT_JULIA_X_CENTER
            yCenter = DEFAULT_JULIA_Y_CENTER
            xExtent = DEFAULT_JULIA_EXTENT
        } else {
            xCenter = DEFAULT_X_CENTER
            yCenter = DEFAULT_Y_CENTER
            xExtent = DEFAULT_X_EXTENT
        }
        numIterations = DEFAULT_ITERATIONS
        initMinMax()
        render()
    }

    fun setColorScheme(colors: IntArray) {
        MandelNative.setColorPalette(paramIndex, colors, colors.size)
    }

    fun setJuliaCoords(jx: Double, jy: Double) {
        this.jx = jx
        this.jy = jy
    }

    fun setCrosshairsEnabled(enabled: Boolean) {
        showCrosshairs = enabled
    }

    @Throws(IOException::class)
    fun savePicture(fileName: String) {
        savePicture(FileOutputStream(fileName))
    }

    @Throws(IOException::class)
    fun savePicture(stream: OutputStream) {
        viewportBitmap!!.compress(Bitmap.CompressFormat.PNG, 100, stream)
        stream.flush()
        stream.close()
    }

    private fun initMinMax() {
        val ratio = screenHeight.toDouble() / (if (screenWidth == 0) 1 else screenWidth).toDouble()
        xmin = xCenter - xExtent / 2.0
        xmax = xCenter + xExtent / 2.0
        ymin = yCenter - ratio * xExtent / 2.0
        ymax = yCenter + ratio * xExtent / 2.0
    }

    private fun getPinchDistance(event: MotionEvent): Float {
        val x = event.getX(0) - event.getX(1)
        val y = event.getY(0) - event.getY(1)
        return sqrt(x * x + y * y.toDouble()).toFloat()
    }

    private fun getPinchCenterPoint(event: MotionEvent, pt: PointF) {
        pt.x = (event.getX(0) + event.getX(1)) * 0.5f
        pt.y = (event.getY(0) + event.getY(1)) * 0.5f
    }

    private inner class MandelThread internal constructor(private val startX: Int, private val startY: Int, private val startWidth: Int, private val startHeight: Int, private val level: Int) : Thread() {
        override fun run() {
            var curLevel = level
            while (true) {
                MandelNative.drawFractal(paramIndex, startX, startY, startWidth, startHeight, curLevel, if (curLevel == level) 1 else 0)
                postInvalidate()
                if (terminateThreads) {
                    break
                }
                if (curLevel <= endCoarseness) {
                    break
                }
                curLevel /= 2
            }
        }
    }

    companion object {
        private const val TAG = "MandelbrotViewBase"
        const val DEFAULT_ITERATIONS = 128
        const val MAX_ITERATIONS = 2048
        const val MIN_ITERATIONS = 2
        const val DEFAULT_X_CENTER = -0.5
        const val DEFAULT_Y_CENTER = 0.0
        const val DEFAULT_X_EXTENT = 3.0
        const val DEFAULT_JULIA_X_CENTER = 0.0
        const val DEFAULT_JULIA_Y_CENTER = 0.0
        const val DEFAULT_JULIA_EXTENT = 3.0
        private const val TOUCH_NONE = 0
        private const val TOUCH_ROTATE = 1
        private const val TOUCH_ZOOM = 2
        private const val CROSSHAIR_WIDTH = 16f
    }
}